tests/many_to_one/tests.py

```diff
@@ -656,3 +656,15 @@ def test_add_then_remove_after_prefetch(self):
         self.assertEqual(city.districts.count(), 2)
         city.districts.remove(d2)
         self.assertEqual(city.districts.count(), 1)
+
+    def test_cache_invalidated_on_save(self):
+        """
+        Saving an object with a manually assigned many to one relationship
+        should invalidate the cached object if the pk points to a different
+        object.
+        """
+        # primes the cache
+        self.assertEqual(self.a.reporter, self.r)
+        self.a.reporter_id = self.r2.pk
+        self.a.save()
```

Comment:
I'm not sure if waiting until `save()` is the best approach. Wouldn't you expect this assertion to pass after the `reporter_id` assignment and before the save?
`self.assertEqual(self.a.reporter, self.r2)`

Reply:
@timgraham Ideally yes but there's no descriptor for ``<field>_id ``, I think it would be too costly to add one, the descriptor would need access to the instance and field to manipulate fields cache.